### ISSUE TITLE
[css-variables] custom property names too permissive, require namespacing rules

### ISSUE THREAD

**Issue by aphillips** [w3c/csswg-drafts#7129]
https://www.w3.org/TR/css-variables-1/#defining-variables

> A custom property is any property whose name starts with two dashes (U+002D HYPHEN-MINUS), like [--foo](). The <custom-property-name> production corresponds to this: it’s defined as any [<dashed-ident>](https://www.w3.org/TR/css-values-4/#typedef-dashed-ident) (a valid [identifier](https://www.w3.org/TR/css-syntax-3/#identifier) that starts with two dashes), except -- itself, which is reserved for future use by CSS. [Custom properties](https://www.w3.org/TR/css-variables-1/#custom-property) are solely for use by authors and users; CSS will never give them a meaning beyond what is presented here.

The above text defines the custom property name as "any valid identifier". Tracing that definition back to [CSS Values](https://www.w3.org/TR/css-values-4/#typedef-dashed-ident) and thence to [`ident token`](https://www.w3.org/TR/css-syntax-3/#typedef-ident-token), we find that the name can contain ***any*** Unicode code point > U+0080. This includes various special forms of whitespace as well as potential problem characters, such as bidi controls (such as might cause "Trojan Source" attacks). Namespacing is definitely a complicated problem: the I18N WG doesn't want groups to cherry-pick characters (thereby excluding certain languages from using the feature).

Most programming languages attempt to address this by adopting some form of restriction for variable names such as those found in [UAX31](https://unicode.org/reports/tr31/) _Unicode Identifier and Pattern Syntax_. In JavaScript, for example, the definition looks like the one found [here](http://es5.github.io/x7.html#x7.6). CSS should make similar restrictions on property _names_ (values can remain unrestricted).

**Comment by tabatkins** [comment]
Yes, custom property names can contain literally any non-ASCII character.

If necessary, I'm happy to restrict this, but I question why it would be necessary to do this for property names, but okay for property values to still be fully unrestricted?

**Comment by aphillips** [comment]
Property names are used in CSS "code" and have to be be parsed, matched, and otherwise referenced. Abusive names can cause spoofing problems (even though the underlying code point sequence is still just some integers and the parser may not care). For example, is `--\0301` a variable reference? Or an error? (using `U+0301 COMBINING ACUTE ACCENT` as an example of a combining mark at the start of a name)

Property values are data and can include natural language text (as well as, well, any character data, including junk). While the value space might be limited by applications in different ways, there don't appear to be any requirements to do so here. In fact, your Spec goes out of its way to highlight this fact:

> Because custom properties can contain anything, there is no general way to know how to interpret what’s inside of them (until they’re substituted into a known property with [var()](https://www.w3.org/TR/css-variables-1/#funcdef-var)). Rather than have them partially resolve in some cases but not others, they’re left completely unresolved; they’re a bare stream of [CSS tokens](https://www.w3.org/TR/css-syntax-3/#tokenization) interspersed with var() functions.

**Comment by tabatkins** [comment]
Property values are used in CSS just as much as property names, tho. We can't interpret custom property values *in the custom property*, but we do as soon as they're substituted into a non-custom property, or if the custom property is registered with a grammar.

Moreso, tho, the `<custom-ident>` value type, which operates on the same rules, is used in a number of places in CSS, such as counter-style names, font names, etc. which seem to be in a similar semantic space.

**Comment by aphillips** [comment]
Yes, it occurred to me that this might turn out to be a gap in CSS Syntax (which might be a serious "ouch" and difficult to do something about).

Since one of the things the property value can contain is a string literal, one probably can't apply UAX31-like rules just generically to the value (i.e. in CSS Variables)

**Comment by tabatkins** [comment]
Right, removing the potential from strings def seems out (at minimum, they def shouldn't be restricted to the Identifier production from Unicode), but I think I didn't make my original point clearly enough - this restriction should apply to all custom identifiers, not just property names, right?

**Comment by aphillips** [comment]
That's right.

**Comment by tabatkins** [comment]
Okay, I'm gonna retag this to Syntax, then, because we should handle the restriction at that level.

**Comment by tabatkins** [comment]
Agenda+ to discuss restricting the allowed codepoints in an [ident sequence](https://drafts.csswg.org/css-syntax/#consume-name) (used in keywords, function names, dimension units, selectors, property names, etc).

Possible options:
* use the [UAX31 categories](https://unicode.org/reports/tr31/#Figure_Code_Point_Categories_for_Identifier_Parsing), matching JS identifiers
* use the post-ASCII part of HTML's [custom element name restrictions](https://html.spec.whatwg.org/multipage/custom-elements.html#valid-custom-element-name), to ensure that selectors can match all custom element names without needing escaping

Then there are subsequent questions. First, what should we do about characters so restricted that *are* used in an identifier sequence?
* Treat them like we do lone surrogates, and replace with U+FFFD
* Disallow them from the production at all, so an identifier valid today with a restricted character in the middle would instead become two identifiers with a DELIM token containing the restricted character. (I'm inclined toward this, as it would cause most usage of the restricted character to become invalid at parse-time, such as in custom property names, and thus would discourage its usage.)

Second, should we allow escapes to represent the restricted character in identifiers? 
* JS, [as far as I can tell](http://es5.github.io/x7.html#x7.6), doesn't allow it. Their restriction is rather broad, tho - it's intended to make it so that *no* identifier that would be illegal to write literally can be written with escapes (in other words, they disallow an ident that could only be written by using escapes) - this disallows things like escaping a dash or period, which CSS has historically allowed and probably can't restrict
* Just let it work. I'm inclined to go this way.

**Comment by svgeesus** [comment]
Since this is now being solved at the CSS Syntax level, which is the correct way to do it, untagging from CSS Variables

**Comment by tabatkins** [comment]
Note that the latter two of the three identifiers in [my popular tweet] would be invalid under the JS rules:

> `.ಠ_ಠ { --（╯°□°）╯: ︵┻━┻; }` is valid CSS.

I'f I'm reading correctly, under the HTML rules the middle one `--（╯°□°）╯` would be invalid due to the degree symbols, but the other two are valid.

Not that this is required to be supported, just noting the effects. ^_^

**Comment by tabatkins** [comment]
The HTML allowed chars are:

> "-" | "." | [0-9] | "_" | [a-z] | #xB7 | [#xC0-#xD6] | [#xD8-#xF6] | [#xF8-#x37D] | [#x37F-#x1FFF] | [#x200C-#x200D] | [#x203F-#x2040] | [#x2070-#x218F] | [#x2C00-#x2FEF] | [#x3001-#xD7FF] | [#xF900-#xFDCF] | [#xFDF0-#xFFFD] | [#x10000-#xEFFFF]

We'd continue to disallow `.` and allow `[A-Z]`, of course, but for all the characters >= 0x80 we'll match this list.

**Comment by svgeesus** [comment]
@aphillips does that resolution address your concerns?

**Comment by aphillips** [comment]
Programming languages such as JS and Java that allow non-ASCII variable names with character limits usually have different restrictions for the initial character. Most notably they forbid combining marks. They sometimes exclude other values (such as bidi controls, although those are excluded above). I think using the HTML restrictions is a realistic solution for CSS for the reasons given above. 

We might need a note about combining mark handling at the start of a token. HTML handles this by requiring an alpha char (`[a-zA-Z]` in your case) at the start (HTML also treats combining characters as non-combining when parsing, e.g. `class="&#x301;"` contains a class name consisting solely of a combining mark. (I used an entity for visibility). As long as the processor can't be fooled, I think you're good to go?

**Comment by tabatkins** [comment]
Ah, hm, indeed. HTML gets to avoid the first-character problem. CSS *does* have special rules for the first character of an ident as well, but they're only different in the ASCII range (preventing idents from being number-lookalikes).

But if the concern is just about combining characters at the start, that'll be fine mechanically; the CSS parser still just handles codepoints individually and doesn't care about combining in any way. So you could select the class you mention without worry, by putting that combining char after a period.

That said, I'm fine with further first-char restrictions if necessary. Unless you request otherwise tho, @aphillips , I'll assume that the existing rules should be fine and use the same non-ASCII allowed characters for both initial and non-initial chars.

**Comment by aphillips** [comment]
As long as the CSS parser doesn't care, then things should work. Content authors will need to be advised about the spoof/abuse potential when viewing CSS files as text somewhere (you may even already have such a note??)

**Comment by tabatkins** [comment]
Don't have such a note, but I'm look over some of the verbage used elsewhere for that issue and add one.

**Comment by tabatkins** [comment]
Okay, restriction applied, and I added a significant note to that section.

I'll need to add and/or tweak some tests for this.

**Comment by mathiasbynens** [comment]
CC’ing @markusicu @macchiati (Unicode “Trojan Source” working group) as FYI

**Comment by tabatkins** [comment]
Note that I'm currently linking to a Rust-lang blog post about the trojan source thing; if there's a better "official source" about it from Unicode I'd be happy to switch the reference over.

**Comment by aphillips** [comment]
Here's [Unicode's announcement](https://home.unicode.org/avoiding-source-code-spoofing/), which has a link to @macchiati et al's doc about the topic.

**Comment by faceless2** [comment]
Re testcases, we've just patched this in and are seeing changes with some old tests:

[css/CSS2/i18n/syndata/character-encoding-001.xht](https://github.com/web-platform-tests/wpt/blob/master/css/CSS2/i18n/syndata/character-encoding-001.xht)
[css/CSS2/i18n/syndata/character-encoding-004.xht](https://github.com/web-platform-tests/wpt/blob/master/css/CSS2/i18n/syndata/character-encoding-004.xht)
[css/CSS2/i18n/syndata/character-encoding-007.xht](https://github.com/web-platform-tests/wpt/blob/master/css/CSS2/i18n/syndata/character-encoding-007.xht)
[css/CSS2/i18n/syndata/character-encoding-010.xht](https://github.com/web-platform-tests/wpt/blob/master/css/CSS2/i18n/syndata/character-encoding-010.xht)
[css/CSS2/i18n/syndata/character-encoding-026.xht](https://github.com/web-platform-tests/wpt/blob/master/css/CSS2/i18n/syndata/character-encoding-026.xht)
[css/CSS2/i18n/syndata/character-encoding-027.xht](https://github.com/web-platform-tests/wpt/blob/master/css/CSS2/i18n/syndata/character-encoding-027.xht)
[css/CSS2/i18n/syndata/character-encoding-028.xht](https://github.com/web-platform-tests/wpt/blob/master/css/CSS2/i18n/syndata/character-encoding-028.xht)
[css/CSS2/i18n/syndata/character-encoding-029.xht](https://github.com/web-platform-tests/wpt/blob/master/css/CSS2/i18n/syndata/character-encoding-029.xht)
[css/CSS2/syntax/characters-0080-009F-001.xht](https://github.com/web-platform-tests/wpt/blob/master/css/CSS2/syntax/characters-0080-009F-001.xht)

**Comment by heilong888855** [comment]
![Screenshot_20220425_022340_com huawei hwid](https://user-images.githubusercontent.com/97087339/167466364-3680d914-cce1-410b-a263-db9a517562b5.jpg)

**Comment by tabatkins** [comment]
Yes, it should be.

**Comment by romainmenke** [comment]
Is it correct that the ranges are inclusive?
Other definitions are explicit about this.

`lowercase letter` has this definition :

> A [code point](https://infra.spec.whatwg.org/#code-point) between U+0061 LATIN SMALL LETTER A (a) and U+007A LATIN SMALL LETTER Z (z) inclusive.

Can we change the current text to :

> non-ASCII ident code point
> A [code point](https://infra.spec.whatwg.org/#code-point) whose value is any of:
> U+00B7
> between U+00C0 and U+00D6 inclusive
> between U+00D8 and U+00F6 inclusive
> between U+00F8 and U+037D inclusive
> between U+037F and U+1FFF inclusive
> U+200C
> U+200D
> U+203F
> U+2040
> between U+2070 and U+218F inclusive
> between U+2C00 and U+2FEF inclusive
> between U+3001 and U+D7FF inclusive
> between U+F900 and U+FDCF inclusive
> between U+FDF0 and U+FFFD inclusive
> greater than or equal to U+10000

**Comment by cdoublev** [comment]
These [tests](https://github.com/web-platform-tests/wpt/blob/188b6f78fbbbf7bf21f2b77cce3bf94621843b5b/dom/nodes/ParentNode-querySelector-escapes.html) also do not appear to have been updated:

```js
testMatched("\u{2003}", "#\u{2003}");
testMatched("\u{A0}", "#\u{A0}");
testMatched("\u{12345}", "#\u{12345}");
```

But is there any intention to revert the change (f1792dd) for this issue? Because HTML now allows more characters in custom element names (`[a-z][^\0\t\n\f\r\u0020/>]*`):

  - https://html.spec.whatwg.org/multipage/custom-elements.html#valid-custom-element-name
  - https://github.com/whatwg/html/pull/7991
